src/main/java/com/google/sps/servlets/FindLocationServlet.java

```diff
@@ -19,7 +27,25 @@
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     
-    //TODO - implement the rest of this function (@Kaleab)
+    String correlator = request.getParameter("correlator");
+
+    //TODO: find a more graceful way to handle this
+    if (correlator == null){
+      throw new IOException();
+    }
+    
+    Filter correlatorFilter =  new FilterPredicate("correlator", FilterOperator.EQUAL, correlator);
+    Query impressionQuery = new Query("Impressions").setFilter(correlatorFilter);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```

Comment:
Feel free to break this into a separate function here and elsewhere.  Took me a while to figure this all out.  A much more readable function could look like this:

public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
  if (!validRequest(request)) {
   // If something other than the creative calls this service return invalid.
   response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
   return;
  }
  storeData(request);
  response.setStatus(HttpServletResponse.SC_OK); 
}

/**
  * Modifies the DB for this ad to show that the user had clicked on "find nearest location"
  * on the ad.
  */
private storeData(HttpServletRequest request) {
  ...
}

Reply:
Done! I structured it this way now